R/pivot-wide.R

```diff
@@ -608,6 +620,52 @@ stop_id_cols_oob <- function(i, arg, call) {
   )
 }
 
+compat_id_cols <- function(id_cols,
+                           ...,
+                           fn_call,
+                           error_call = caller_env(),
+                           user_env = caller_env(2)) {
+  dots <- enquos(...)
+
+  # If `id_cols` is specified by name by the user, it will show up in the call.
+  # Otherwise, default args don't show up in the call so it won't be there.
+  user_specified_id_cols <- "id_cols" %in% names(fn_call)
+
+  # For compatibility (#1353), assign the first value of `...` to `id_cols` if:
+  # - The user didn't specify `id_cols`.
+  # - There is exactly 1 unnamed element in `...`.
+  use_compat_id_cols <-
+    !user_specified_id_cols &&
+    length(dots) == 1L &&
+    !is_named(dots)
+
+  if (use_compat_id_cols) {
+    id_cols <- dots[[1L]]
+    warn_deprecated_unnamed_id_cols(id_cols, user_env = user_env)
+  } else {
+    id_cols <- enquo(id_cols)
+    check_dots_empty0(..., call = error_call)
+  }
+
+  id_cols
+}
+
+warn_deprecated_unnamed_id_cols <- function(id_cols, user_env = caller_env(2)) {
+  id_cols <- as_label(id_cols)
+
+  lifecycle::deprecate_warn(
+    when = "1.3.0",
+    what = I(cli::format_inline(
+      "Specifying the {.arg id_cols} argument by position"
+    )),
+    details = cli::format_inline(
+      "Please explicitly name {.arg id_cols}, like {.code id_cols = {id_cols}}."
+    ),
+    always = TRUE,
```

Comment:
Fairly aggressive about this compatibility deprecation warning, which felt alright to me since it is an easy fix on the user side and we'd like to remove this as soon as we reasonably can